src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -369,8 +369,18 @@ public unsafe int GetLowerBound(int dimension)
             return Unsafe.Add(ref RuntimeHelpers.GetMultiDimensionalArrayBounds(this), rank + dimension);
         }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal extern CorElementType GetCorElementTypeOfElementType();
+        internal unsafe CorElementType GetCorElementTypeOfElementType()
```

Comment:
This extra layer will add measurable overhead. For example, try to benchmark

```
   int[] a = new int[10];
   for(;;)
   {
       int start = Environment.TickCount;
       for (int i = 0; i < 100000000; i++) Buffer.ByteLength(a);
       int end = Environment.TickCount;
       Console.WriteLine((end-start).ToString());
    } 
```

I am seeing ~400ms before this change and ~500ms after this change.

Reply:
Maybe inlining issue? Try JITDump to see if it's inlined?
I'm going to identify the codegen here.

Reply:
For this specific case, it can be optimized further: no need to go through the EEClass to acquire the CorElementType. Just check the MethodTable to see if primitive flag is set.
All the usages of `IsPrimitiveType` can actually benefit from this.
A question: in which cases, the element of array will be `TypeDesc` instead of `MethodTable`?

Reply:
Adding `AggressiveInlining` mitigates this regression:
|                Method |        Job |                       Toolchain |      Mean |     Error |    StdDev | Ratio | RatioSD |
|---------------------- |----------- |-------------------------------- |----------:|----------:|----------:|------:|--------:|
|       ArrayIndexOfInt | Job-SOMQPI | \FCallAggressiveInl\corerun.exe | 15.128 ns | 0.0553 ns | 0.0462 ns |  1.02 |    0.00 |
|       ArrayIndexOfInt | Job-ACSXKB |              \FCall\corerun.exe | 15.787 ns | 0.0558 ns | 0.0494 ns |  1.06 |    0.01 |
|       ArrayIndexOfInt | Job-MHGPKF |               \main\corerun.exe | 14.898 ns | 0.0638 ns | 0.0566 ns |  1.00 |    0.00 |
|                       |            |                                 |           |           |           |       |         |
|      ArrayIndexOfEnum | Job-SOMQPI | \FCallAggressiveInl\corerun.exe | 15.351 ns | 0.0998 ns | 0.0933 ns |  0.98 |    0.01 |
|      ArrayIndexOfEnum | Job-ACSXKB |              \FCall\corerun.exe | 15.987 ns | 0.1452 ns | 0.1287 ns |  1.02 |    0.01 |
|      ArrayIndexOfEnum | Job-MHGPKF |               \main\corerun.exe | 15.609 ns | 0.1262 ns | 0.1054 ns |  1.00 |    0.00 |
|                       |            |                                 |           |           |           |       |         |
|     ArrayIndexOfClass | Job-SOMQPI | \FCallAggressiveInl\corerun.exe |  6.921 ns | 0.0355 ns | 0.0296 ns |  1.02 |    0.00 |
|     ArrayIndexOfClass | Job-ACSXKB |              \FCall\corerun.exe |  6.921 ns | 0.0344 ns | 0.0305 ns |  1.02 |    0.01 |
|     ArrayIndexOfClass | Job-MHGPKF |               \main\corerun.exe |  6.782 ns | 0.0265 ns | 0.0221 ns |  1.00 |    0.00 |
|                       |            |                                 |           |           |           |       |         |
| ArrayIndexOfValueType | Job-SOMQPI | \FCallAggressiveInl\corerun.exe | 10.622 ns | 0.0471 ns | 0.0394 ns |  1.00 |    0.00 |
| ArrayIndexOfValueType | Job-ACSXKB |              \FCall\corerun.exe | 11.532 ns | 0.0687 ns | 0.0643 ns |  1.09 |    0.01 |
| ArrayIndexOfValueType | Job-MHGPKF |               \main\corerun.exe | 10.622 ns | 0.0489 ns | 0.0433 ns |  1.00 |    0.00 |
|                       |            |                                 |           |           |           |       |         |
|            ByteLength | Job-SOMQPI | \FCallAggressiveInl\corerun.exe |  3.316 ns | 0.0174 ns | 0.0163 ns |  1.06 |    0.01 |
|            ByteLength | Job-ACSXKB |              \FCall\corerun.exe |  4.412 ns | 0.0188 ns | 0.0176 ns |  1.41 |    0.02 |
|            ByteLength | Job-MHGPKF |               \main\corerun.exe |  3.134 ns | 0.0458 ns | 0.0406 ns |  1.00 |    0.00 |